dev/codeowners/codeowners.go

```diff
@@ -24,28 +25,88 @@ func Check() error {
 	if err != nil {
 		return err
 	}
-
 	const packagesDir = "packages"
-	return filepath.WalkDir(packagesDir, func(path string, d fs.DirEntry, err error) error {
-		if d.IsDir() {
-			if path != packagesDir && filepath.Dir(path) != packagesDir {
-				return fs.SkipDir
-			}
-			return nil
-		}
-		if d.Name() != "manifest.yml" {
-			return nil
-		}
+	if err := validatePackages(codeowners, packagesDir); err != nil {
+		return err
+	}
 
-		return codeowners.checkManifest(path)
-	})
+	return nil
 }
 
 type githubOwners struct {
 	owners map[string][]string
 	path   string
 }
 
+// validatePackages checks if all packages in packagesDir have a manifest.yml file
+// with the correct owner as captured in codeowners. Also, for packages that share ownership across
+// data_streams, it checks that all data_streams are explicitly owned by a single owner. Such ownership
+// sharing packages are identified by having at least one data_stream with explicit ownership in codeowners.
+func validatePackages(codeowners *githubOwners, packagesDir string) error {
+	packageDirEntries, err := os.ReadDir(packagesDir)
+	if err != nil {
+		return err
+	}
+
+	if len(packageDirEntries) == 0 {
+		if len(codeowners.owners) == 0 {
+			return nil
+		}
+		return fmt.Errorf("no packages found in %q", packagesDir)
+	}
+
+	for _, packageDirEntry := range packageDirEntries {
+		packageName := packageDirEntry.Name()
+
+		packagePath := path.Join(packagesDir, packageName)
+
+		packageManifestPath := path.Join(packagePath, "manifest.yml")
+		err = codeowners.checkManifest(packageManifestPath)
+		if err != nil {
+			return err
+		}
+
+		packageDataStreamsPath := path.Join(packagePath, "data_stream")
+		if _, err := os.Stat(packageDataStreamsPath); os.IsNotExist(err) {
+			// package doesn't have data_streams
+			continue
+		}
+
+		dataStreamDirEntries, err := os.ReadDir(packageDataStreamsPath)
+		if err != nil {
+			return err
+		}
+
+		totalDataStreams := len(dataStreamDirEntries)
+		if totalDataStreams == 0 {
+			// package doesn't have data_streams
+			continue
+		}
+
+		var dataStreamsWithoutOwner []string
+		for _, dataStreamDirEntry := range dataStreamDirEntries {
+			dataStreamName := dataStreamDirEntry.Name()
+			dataStreamDir := path.Join(packageDataStreamsPath, dataStreamName)
+			dataStreamOwners, found := codeowners.owners["/"+dataStreamDir]
+			if !found {
+				dataStreamsWithoutOwner = append(dataStreamsWithoutOwner, dataStreamDir)
+				continue
+			}
+			if len(dataStreamOwners) > 1 {
+				return errors.New(fmt.Sprintf("data stream \"%s\" of package \"%s\" has more than one owners [%s]", dataStreamDir,
+					packagePath, strings.Join(dataStreamOwners, ", ")))
```

Comment:
Nit. Use `fmt.Errorf` directly.

```suggestion
				return fmt.Errorf("data stream \"%s\" of package \"%s\" has more than one owners [%s]", dataStreamDir,
					packagePath, strings.Join(dataStreamOwners, ", "))
```